app/services/user_event_creator.rb

```diff
@@ -77,19 +74,22 @@ def build_disavowal_token
   # @return [Array(Event, String)] an (event, disavowal_token) tuple
   def create_event_for_new_device(event_type:, user:, disavowal_token:)
     if user.fully_registered? && user.has_devices? && disavowal_token.nil?
-      device, event, disavowal_token = Device.transaction do
-        device = create_device_for_user(user)
-        event, disavowal_token = create_user_event_with_disavowal(
-          event_type, user, device
+      if event_type != :sign_in_before_2fa
```

Comment:
Am I reading this correctly that we would no longer be sending the `:sign_in_before_2fa` at all? I think we still want to send that device notification.

I see the challenge with the user not being considered "new" anymore by the time they reach the MFA screen. To work around that, maybe we can assign a value in the session before we call `create_user_event(:sign_in_before_2fa)` indicating whether they were a new device at sign-in, and reference that session value from the MFA analytics.

Reply:
I have it so we are still sending it but after the MFA is submitted. If I'm understanding the workflow correctly, the way it currently works is a user lacking the device cookie enters their email/password and as they're sent to the MFA (sign_in_before_2fa) the notification is sent along with the user_event being created. Since it happens before MFA is submitted it precedes the MFA analytics event being recorded.

By omitting the sign_in_before type it causes the create event for new device to skip and run when sign_in_after type happens. When sign_in_after type hits the create_event function it goes ahead and sends the notification... 

It sounds like if we assign a session value and catch it in analytics this might be less complicated of a solution, right?

Reply:
I moved that check into user session and updated downstream. That puts user_event_creator back the way it was before.